fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java

```diff
@@ -104,4 +117,79 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
+        LocalDateTime bottomLocalDateTime = ((ConstantOperator) child2).getDatetime().toLocalDate().atTime(0, 0, 0, 0);
+        LocalDateTime targetLocalDateTime;
+        BinaryPredicateOperator.BinaryType binaryType = operator.getBinaryType();
+        int offset;
+        if (binaryType.equals(BinaryPredicateOperator.BinaryType.GE)) {
+            // when the BinaryType is >= ,cast dateTime to minimum date type；
+            // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-28)
+            // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-29)
+            if (originalDateTime.isEqual(bottomLocalDateTime)) {
+                offset = 0;
+            } else {
+                offset = 1;
+            }
+            targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
+            ConstantOperator newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+            BinaryPredicateOperator binaryPredicateOperator =
+                    new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.GE, castChild, newDate);
+            return binaryPredicateOperator;
+        } else if (binaryType.equals(BinaryPredicateOperator.BinaryType.GT)) {
+            // when the BinaryType is > ,cast dateTime to minimum date type；
+            // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-29)
+            // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-29)
+            offset = 1;
+            targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
+            ConstantOperator newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+            BinaryPredicateOperator binaryPredicateOperator =
+                    new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.GE, castChild, newDate);
+            return binaryPredicateOperator;
+        } else if (binaryType.equals(BinaryPredicateOperator.BinaryType.LE)) {
+            // when the BinaryType is <= ,cast dateTime to maximum date type；
+            // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-28)
+            // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-28)
+            offset = 0;
+            targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
+            ConstantOperator newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+            BinaryPredicateOperator binaryPredicateOperator =
+                    new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, castChild, newDate);
+            return binaryPredicateOperator;
+        } else if (binaryType.equals(BinaryPredicateOperator.BinaryType.LT)) {
+            // when the BinaryType is < ,cast dateTime to maximum date type；
+            // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-27)
+            // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-28)
```

Comment:
This offset is -1,  dateTime(2021-12-28 00:00:00.1) will cast to date(2021-12-27).